### PR TITLE
debug: allow a single debug extension to provide multiple configs

### PR DESCRIPTION
**Pull request by connor4312**
Previously if a debug extension provided multiple dynamic
configurations, we would just use the first debugger -- whatever that
was. This change now shows all configurations for which dynamic configs
are registered.

I also adjusted the picker to automatically select the first item if
there's only a single configuration provided. This works well for the
debug terminal, but also means that the user doesn't see the name of
the selected item, which might not be desirable. Open to pushback.

Together these finish the request for a separate top-level contribution
for the terminal in  #98054

Finally, with that adjustment I made a tweak so that the picker shows
up in a `busy` state while extensions are activating. Previously you
would select a dynamic configuration title and could have a few seconds
of delay before the picker came up, which is probably not desirable.

### CODE REVIEW

File: src/vs/workbench/contrib/debug/browser/debugConfigurationManager.ts
Comment (isidorn) [inline src/vs/workbench/contrib/debug/browser/debugConfigurationManager.ts:263]: Why not `return acc.concat(explicitTypes)`, for me it reads better

File: src/vs/workbench/contrib/debug/browser/debugConfigurationManager.ts
Comment (isidorn) [inline src/vs/workbench/contrib/debug/browser/debugConfigurationManager.ts:266]: It feels a bit clunky that we also have this check after the check above. Is it possible to combine these two into one, and just have one code path.
Maybe not, just a thought.

File: src/vs/workbench/contrib/debug/browser/debugConfigurationManager.ts
Comment (isidorn) [inline src/vs/workbench/contrib/debug/browser/debugConfigurationManager.ts:315]: Not a fan of combining `await` and `then`. Just does not make it very readable for me. Can we get rid of the `then`

File: src/vs/workbench/contrib/debug/browser/debugConfigurationManager.ts
Comment (isidorn) [inline src/vs/workbench/contrib/debug/browser/debugConfigurationManager.ts:285]: Feels like these are all listeners which should be disposed.
I see that you dispose the `input` but maybe it is cleaner if you dispose all the listeners as well.

File: src/vs/workbench/contrib/debug/browser/debugConfigurationManager.ts
Comment (isidorn) [inline src/vs/workbench/contrib/debug/browser/debugConfigurationManager.ts:278]: Yeah I personally dislike the `createQuickPick` API since it brings too much noisy compared to the previous `.pick` usage.
Are we using the new `createQuickPick` only for the `.busy` flag or is ther emore to it

**Review by isidorn — CHANGES_REQUESTED** [review]

File: src/vs/workbench/contrib/debug/browser/debugConfigurationManager.ts
Comment (isidorn) [inline src/vs/workbench/contrib/debug/browser/debugConfigurationManager.ts:317]: Add a one line comment that we automatically pick the first item if there is only one

**Comment by isidorn** [comment]
@connor4312 thanks a lot for this PR. Good work overall!
I did a first review and left comments in code. Once we adress that I can do a second review and test out the flow.

I like the optimisation to not bother the user if there is only one Configuration!

As for PRs in general, I am a big fan and keep em coming. Also feel free to split PRs in multiple if they tackle seperate things since that makes it easier for me to review and merge.
Also might you be interested in improving this https://github.com/microsoft/vscode/issues/96293
Since it is in this code area. I can also look into it once we merge this one in so we do not hit conflicts. Thanks!

File: src/vs/workbench/contrib/debug/browser/debugConfigurationManager.ts
Comment (connor4312) [reply src/vs/workbench/contrib/debug/browser/debugConfigurationManager.ts:263]: I have been influenced too much by all the react hipsters who who use spreads for everything 😄 I will adjust

File: src/vs/workbench/contrib/debug/browser/debugConfigurationManager.ts
Comment (connor4312) [reply src/vs/workbench/contrib/debug/browser/debugConfigurationManager.ts:278]: It's so that we can create the quick pick before we start activating debuggers and getting results, and so that we have control over it to auto-select a single item if there's only one (if we just wrapped everything in a promise given to `.pick` I don't think we'd be able to do that automatic selector)

Comment (isidorn) [reply src/vs/workbench/contrib/debug/browser/debugConfigurationManager.ts:278]: Ok, makes sense

File: src/vs/workbench/contrib/debug/browser/debugConfigurationManager.ts
Comment (isidorn) [reply src/vs/workbench/contrib/debug/browser/debugConfigurationManager.ts:263]: Heheh

**Comment by isidorn** [comment]
Looks great. Merging in.

**Pull request merged**
